File: app/src/main/java/com/rollapp/shared/ui/components/Develop.kt
package com.rollapp.shared.ui.components

import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.tween
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.blur
import androidx.compose.ui.graphics.ColorFilter
import androidx.compose.ui.graphics.ColorMatrix
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import coil.compose.AsyncImage
import coil.request.ImageRequest

/**
 * An image that *develops*: it arrives blurred, desaturated and dim, and resolves to
 * the real photograph over about a second — the way a print comes up in the tray.
 *
 * The animation starts when the bitmap is ready, not when the composable appears, so
 * a slow network shows the placeholder, not a half-developed nothing. Blur needs
 * API 31; older devices get the saturation and brightness ramp alone.
 */
@Composable
fun DevelopingImage(
    model: Any?,
    contentDescription: String?,
    modifier: Modifier = Modifier,
    contentScale: ContentScale = ContentScale.Crop,
    durationMillis: Int = 1100,
    delayMillis: Int = 0
) {
    val progress = remember(model) { Animatable(0f) }
    var ready by remember(model) { mutableStateOf(false) }

    LaunchedEffect(ready) {
        if (ready) progress.animateTo(1f, tween(durationMillis, delayMillis, easing = Settle))
    }

    val p = progress.value
    val filter = remember(p) {
        ColorFilter.colorMatrix(
            ColorMatrix().apply {
                setToSaturation(0.2f + 0.8f * p)
                val b = 0.55f + 0.45f * p
                timesAssign(ColorMatrix(floatArrayOf(
                    b, 0f, 0f, 0f, 0f,
                    0f, b, 0f, 0f, 0f,
                    0f, 0f, b, 0f, 0f,
                    0f, 0f, 0f, 1f, 0f
                )))
            }
        )
    }

    AsyncImage(
        model = ImageRequest.Builder(LocalContext.current)
            .data(model)
            .crossfade(false)
            .build(),
        contentDescription = contentDescription,
        contentScale = contentScale,
        colorFilter = if (p < 1f) filter else null,
        onSuccess = { ready = true },
        modifier = modifier
            .graphicsLayer { alpha = 0.4f + 0.6f * p }
            .blur(((1f - p) * 10).dp)
    )
}

File: ios/WeWere/UI/Components/Develop.swift
import SwiftUI
import Kingfisher

/// An image that *develops*: it arrives blurred, desaturated and dim, and resolves to
/// the real photograph over about a second — the way a print comes up in the tray.
///
/// The animation starts when the bitmap is ready, not when the view appears, so a
/// slow network shows the placeholder, not a half-developed nothing.
struct DevelopingImage: View {
    let url: String?
    var contentMode: SwiftUI.ContentMode = .fill
    var duration: Double = 1.1
    var delay: Double = 0
    /// Longest edge to decode at, for tiles that never need the full 2560px image.
    var downsample: CGFloat? = nil

    @State private var p: CGFloat = 0

    var body: some View {
        let source = url.flatMap(URL.init(string:))
        RemoteImage(url: source, contentMode: contentMode, downsample: downsample, onReady: {
            guard p == 0 else { return }
            withAnimation(.settle(duration: duration, delay: delay)) { p = 1 }
        })
        .saturation(0.2 + 0.8 * p)
        .brightness(-0.22 * (1 - p))
        .opacity(0.4 + 0.6 * p)
        .blur(radius: (1 - p) * 10)
        .id(url)
    }
}

/// Coil's `AsyncImage`: a cached network image that fills or fits its frame.
struct RemoteImage: View {
    let url: URL?
    var contentMode: SwiftUI.ContentMode = .fill
    var downsample: CGFloat? = nil
    var fade: Bool = false
    var onReady: (() -> Void)? = nil

    var body: some View {
        Color.clear.overlay(
            KFImage(url)
                .setProcessors(downsample.map { [DownsamplingImageProcessor(size: CGSize(width: $0, height: $0))] } ?? [])
                .cacheOriginalImage()
                .fade(duration: fade ? 0.25 : 0)
                .onSuccess { _ in onReady?() }
                .resizable()
                .aspectRatio(contentMode: contentMode)
        )
        .clipped()
    }
}

/// A local file (a capture or a staged upload), decoded at a sane size.
struct LocalImage: View {
    let url: URL?
    var contentMode: SwiftUI.ContentMode = .fill
    var downsample: CGFloat = 800

    var body: some View {
        Color.clear.overlay(
            KFImage(url)
                .setProcessor(DownsamplingImageProcessor(size: CGSize(width: downsample, height: downsample)))
                .cacheMemoryOnly()
                .resizable()
                .aspectRatio(contentMode: contentMode)
        )
        .clipped()
    }
}
